docs/architecture/adr-020-protobuf-transaction-encoding.md

```diff
@@ -24,76 +25,186 @@ With this in mind, we will tackle the migration path via two main areas, txs and
-to provide canonical representation of a `Transaction` to sign over, clients must
-obey the following rules:
+Again, it does not matter if `TxBody` was encoded canonically or not.
 
-- Encode `SignDoc` (see below) via [Protobuf's canonical JSON encoding](https://developers.google.com/protocol-buffers/docs/proto3#json).
-  - Default must be stripped from the output!
-  - JSON keys adhere to their Proto-defined field names.
-- Generate canonical JSON to sign via the [JSON Canonical Form Spec](https://gibson042.github.io/canonicaljson-spec/).
-  - This spec should be trivial to interpret and implement in any language.
+Note that in order to decode `SignDocRaw` for displaying contents to users,
+the regular `SignDoc` type should be used.
 
-```Protobuf
-// app/codec/codec.proto
+3. Broadcast `TxRaw`
 
-message SignDoc {
-  StdSignDocBase base = 1;
-  repeated Message msgs = 2;
+In order to make sure that the signed body bytes exactly match the encoded body
+bytes, clients should encode and broadcast `TxRaw` with the same body bytes used
+in `SignDocRaw`.
+
+```proto
+// types/types.proto
+message TxRaw {
+    bytes body_bytes = 1;
+    repeated bytes signatures = 2;
 }
 ```
 
+Signature verifiers should verify signatures by decoding `TxRaw` and then
+encoding `SignDocRaw` with the raw body bytes.
+
+The standard `Tx` type (which is byte compatible with `TxRaw`) can be used to
+decode transactions for all other use cases.
+
+#### `SIGN_MODE_LEGACY_AMINO`
+
+In order to support legacy wallets and exchanges, Amino JSON will be emporarily
+supported transaction signing. Once wallets and exchanges have had a
+chance to upgrade to protobuf based signing, this option will be disabled. In
+the meantime, it is foreseen that disabling the current Amino signing would cause
+too much breakage to be feasible.
+
+Legacy clients will be able to sign a transaction using the current Amino
+JSON format and have it encoded to protobuf using the REST `/tx/encode`
+endpoint before broadcasting.
+
+#### `SIGN_MODE_EXTENDED`
+
+As was discussed extensively in [\#6078](https://github.com/cosmos/cosmos-sdk/issues/6078),
+there is a desire for a human-readable signing encoding, especially for hardware
+wallets like the [Ledger](https://www.ledger.com) which display
+transaction contents to users before signing. JSON was an attempt at this but 
+falls short of the ideal.
+
+`SIGN_MODE_EXTENDED` is intended as a placeholder for a human-readable
+encoding which will replace Amino JSON. This new encoding should be even more
+focused on readability than JSON, possibly based on formatting strings like
+[MessageFormat](http://userguide.icu-project.org/formatparse/messages).
+
+In order to ensure that the new human-readable format does not suffer from
+transaction malleability issues, `SIGN_MODE_EXTENDED`
+requires that the _human-readable bytes are concatenated with the raw `SignDoc`_
+to generate sign bytes.
```

Comment:
We could also include a checksum in the textual representation. That would be easier to print and could also serve as a plausibility check between the tx composing environment and the tx signing environment, which that full content can hardly do.

Reply:
What do you mean by plausibility check?

Reply:
A wallet or website (basically any composing environment) that composes a transaction to be signed can display a checksum of the unsigned transaction like: _Please sign transaction with ID h398gh3894hg038hg08 to prroceed_. Then the signing environment can display all the human readable content along with this identifier (instead of a blob), allowing the user to check that both environments are talking about the same binary data.

Since the checksum changes entirely when only some parts of the binary data change (e.g. data is cut), you would see a difference that you don't see when looking at the blob.

Reply:
Okay, interesting idea. I don't think we need to decide on the exact convention in this document. We should continue that discussion in the forum where we'll be discussing that human readable format. Will ping once a thread is started.